dbms/src/Flash/Mpp/TrackedMppDataPacket.h

```diff
@@ -0,0 +1,192 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <common/logger_useful.h>
+#include <common/types.h>
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
+#ifdef __clang__
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+#include <Common/MemoryTracker.h>
+#include <grpcpp/server_context.h>
+#include <kvproto/mpp.pb.h>
+#include <kvproto/tikvpb.grpc.pb.h>
+#include <tipb/select.pb.h>
+#pragma GCC diagnostic pop
+#include <Common/UnaryCallback.h>
+
+#include <memory>
+
+namespace DB
+{
+inline size_t estimateAllocatedSize(const mpp::MPPDataPacket & data)
+{
+    size_t ret = data.data().size();
+    for (int i = 0; i < data.chunks_size(); i++)
+    {
+        ret += data.chunks(i).size();
+    }
+    return ret;
+}
+
+struct MemTrackerWrapper
+{
+    MemTrackerWrapper(size_t _size, MemoryTracker * memory_tracker)
+        : memory_tracker(memory_tracker)
+        , size(0)
+    {
+        alloc(_size);
+    }
+
+    void alloc(size_t delta)
+    {
+        if (memory_tracker)
+        {
+            memory_tracker->alloc(delta);
+            size += delta;
+        }
+    }
+
+    void switchMemTracker(MemoryTracker * new_memory_tracker)
+    {
+        int bak_size = size;
+        freeAll();
+        memory_tracker = new_memory_tracker;
+        alloc(bak_size);
+    }
+    ~MemTrackerWrapper()
+    {
+        freeAll();
+    }
+
+    void freeAll()
+    {
+        if (size)
+        {
+            if (memory_tracker)
+            {
+                memory_tracker->free(size);
+                size = 0;
+            }
+        }
+    }
+    MemoryTracker * memory_tracker;
+    size_t size = 0;
+};
+
+struct TrackedMppDataPacket
+{
+    explicit TrackedMppDataPacket(const mpp::MPPDataPacket & data, MemoryTracker * memory_tracker)
+        : mem_tracker_wrapper(estimateAllocatedSize(data), memory_tracker)
+    {
+        packet = data;
+    }
+
+    explicit TrackedMppDataPacket()
+        : mem_tracker_wrapper(0, current_memory_tracker)
+    {}
+
+    explicit TrackedMppDataPacket(MemoryTracker * memory_tracker)
+        : mem_tracker_wrapper(0, memory_tracker)
+    {}
+
+    void addChunk(std::string && value)
+    {
+        mem_tracker_wrapper.alloc(value.size());
+        packet.add_chunks(value);
+    }
+
+    void serializeByResponse(const tipb::SelectResponse & response)
+    {
+        mem_tracker_wrapper.alloc(response.ByteSizeLong());
+        if (!response.SerializeToString(packet.mutable_data()))
+            throw Exception(fmt::format("Fail to serialize response, response size: {}", response.ByteSizeLong()));
```

Comment:
need to free the memory allocated in L118?

Reply:
good spot

Reply:
fixed